docs/source/ddp_comm_hooks.rst

```diff
@@ -0,0 +1,64 @@
+DDP Communication Hooks
+=======================
+
+DDP communication hook is a generic interface to control how to communicate
+gradients across workers.
+A few built-in communication hooks are provided,
+and users can easily apply any of these hooks to optimize communication.
+Besides, the hook interface can also support user-defined communication
+strategies for more advanced use cases.
+
+.. warning ::
+    DDP communication hooks can only support single process single device mode
+    on NCCL backend.
+
+How to Use A Communication Hook?
+--------------------------------
+
+To use a communication hook, the user just needs to let the DDP model register
+the hook before the training loop.
+
+.. automethod:: torch.nn.parallel.DistributedDataParallel.register_comm_hook
+
+Default Communication Hooks
+---------------------------
+
+Default communication hooks are simple **stateless** hooks, so the input state
+in ``register_comm_hook`` is either a process group or ``None``.
+
+.. automodule:: torch.distributed.algorithms.ddp_comm_hooks.default_hooks
+    :members:
+
+PowerSGD Communication Hook
+---------------------------
+
+PowerSGD communication hook is a **stateful** hook used for gradient
+compression, and the user needs to provide a state defined as below.
+The performance is `on par with <https://observablehq.com/@tvogels/powersgd-benchmark>`_
+the implementation in the original `paper <https://arxiv.org/abs/1905.13727>`_.
+
+PowerSGD State
+^^^^^^^^^^^^^^^^
+
+.. currentmodule:: torch.distributed.algorithms.ddp_comm_hooks.powerSGD_hook
+.. autoclass:: PowerSGDState
+
+PowerSGD Hooks
```

Comment:
Maybe Add a note that there are batched and unbatched versions, as well as the tradeoffs between them?

Reply:
Already covered by the docstrings. Copied it here:

This variant is faster than powerSGD_hook() that runs layer-wise gradient compression, but it usually results in a much lower accuracy, unless matrix_approximation_rank in the state is 1. Increasing matrix_approximation_rank may not necessarily increase the accuracy, because batching per-parameter tensors without column/row alignment can destroy low-rank structure. Therefore, the user should always consider powerSGD_hook() first, and only consider this variant when a satisfying accuracy can be achieved when matrix_approximation_rank is 1.